### PR TITLE
Extract "webhooks & events" to separate page

### PR DESCRIPTION
**Pull request by orzechdev**
It introduces the following:
- extracted custom apps with webhook and events out of the rest external apps,
- moved the list of custom apps and details views from `/apps/` paths to `/custom-apps/` and refactored routing.

Closes https://github.com/saleor/saleor-app-marketplace/issues/133

**PR intended to be tested with API branch:** 

### Screenshots

Screenshot 2022-12-13 at 16 26 31

### Pull Request Checklist

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `[data-test-id]` are added for new elements
4. [ ] Changes are mentioned in the changelog
5. [ ] The changes are tested in different browsers and in light/dark mode

### Test environment config

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
MARKETPLACE_URL=https://apps.saleor.io
SALEOR_APPS_ENDPOINT=https://apps.saleor.io/api/saleor-apps

### Do you want to run more stable tests?
To run all tests, just select the stable checkbox. To speed up tests, increase the number of containers. Tests will be re-run only when the "run e2e" label is added.

1. [ ] stable
2. [ ] giftCard
3. [ ] category
4. [ ] collection
5. [ ] attribute
6. [ ] productType
7. [ ] shipping
8. [ ] customer
9. [ ] permissions
10. [ ] menuNavigation
11. [ ] pages
12. [ ] sales
13. [ ] vouchers
14. [ ] homePage
15. [ ] login
16. [ ] orders
17. [ ] products
18. [x] app
19. [ ] plugins
20. [ ] translations
21. [ ] navigation
22. [ ] variants
23. [ ] payments

CONTAINERS=1

### CODE REVIEW

**Review by lkostrowski — DISMISSED** [review]

File: src/custom-apps/index.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  customAppWebhookPath,
+} from "./urls";
+import CustomAppCreateView from "./views/CustomAppCreate";
+import CustomAppDetailsView from "./views/CustomAppDetails";
+import CustomAppListView from "./views/CustomAppList";
+import CustomAppWebhookCreateView from "./views/CustomAppWebhookCreate";
+import CustomAppWebhookDetailsView from "./views/CustomAppWebhookDetails";
+
+const CustomAppList: React.FC<RouteComponentProps> = () => {
+  const qs = parseQs(location.search.substr(1));
```
[the remark this answers is not in the document]
Comment (Droniu) [reply src/custom-apps/index.tsx:24]: It's a good idea, but I believe we should do refactor it everywhere at once. I've made an issue for this: #2836

File: src/custom-apps/views/CustomAppList.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import {
+  customAppListUrl,
+  CustomAppListUrlDialog,
+  CustomAppListUrlQueryParams,
+  customAppUrl,
+} from "../urls";
+
+const getCurrentAppName = (
+  id: string,
+  collection?: AppsListQuery["apps"]["edges"],
```
Comment (Droniu) [inline src/custom-apps/views/CustomAppList.tsx:31]: **suggestion:** We probably have a fragment for that, something like AppFragment.

File: src/custom-apps/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrzejewsky) [reply src/custom-apps/index.tsx:24]: fair point, we want to migrate from `qa` anyway

File: src/custom-apps/views/CustomAppList.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  return (
+    <>
+      <WindowTitle title={intl.formatMessage(sectionNames.webhooksAndEvents)} />
+      <AppDeleteDialog
+        confirmButtonState={deleteAppOpts.status}
+        name={getCurrentAppName(params.id, customApps)}
+        onClose={closeModal}
+        onConfirm={handleRemoveConfirm}
+        type={"CUSTOM"}
```
Comment (Droniu) [inline src/custom-apps/views/CustomAppList.tsx:110]:
```suggestion
        type="CUSTOM"
```

**Review by Droniu — APPROVED** [review]

File: src/custom-apps/urls.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+export const customAppWebhookAddPath = (appId: string) =>
+  urlJoin(customAppsSection, appId, webhooksSection, "add");
+
+export const customAppWebhookAddUrl = (
+  appId: string,
+  params?: CustomAppDetailsUrlQueryParams,
+) =>
+  customAppWebhookAddPath(encodeURIComponent(appId)) +
+  "?" +
+  stringifyQs(params);
```
[the remark this answers is not in the document]
Comment (andrzejewsky) [reply src/custom-apps/urls.ts:57]: +1 we need to do similar thing with our config

File: src/custom-apps/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (lkostrowski) [reply src/custom-apps/index.tsx:24]: Ok, so lets leave it like it is in this PR @orzechdev

**Review by andrzejewsky — CHANGES_REQUESTED** [review]
Please write unit tests (it's part of our DoD), at least for the components/code you have created. Please think also about testing the existing code you amended (at least critical path)

**Comment by lkostrowski** [comment]
@andrzejewsky AFAIK this PR doesn't have new features, it extracted part of the apps page to a separate page. I agree to write unit tests (as always) but let's not block Marketplace release by writing tests to legacy code that already worked before. I believe proper testing will require a huge refactor which can be a separate PR (to be safe not mix tests with functionality)

**Comment by andrzejewsky** [comment]
@lkostrowski this is part of our legacy management, writing units even for the legacy code is very important for us. Yeah, it worked, but we have no guarantee it will continue to work. It's not really recommended way to skip tests and treat them as a blocker, it's rather part of the implementation. I'm asking for simple units just critical stuff, I don't expect 100% coverage.

If that's really urgent, let's skip it. However, I really ask for including writing tests in the estimations (at least simple ones)

**Review by andrzejewsky — APPROVED** [review]

**Comment by lkostrowski** [comment]
> @lkostrowski this is part of our legacy management, writing units even for the legacy code is very important for us. Yeah, it worked, but we have no guarantee it will continue to work. It's not really recommended way to skip tests and treat them as a blocker, it's rather part of the implementation. I'm asking for simple units, for just critical stuff, I don't expect 100% coverage.
> 
> If that's really urgent, let's skip it. However, I really ask for including writing tests in the estimations (at least simple ones)

I agree, I'm just not sure if unit tests are the best

From my experience legacy code is easier to test with higher-level tests (observable behavior, integrations), while units are best when code is refactored/extracted. 

I'm just worried that code is not testable and we will spend days on mocking or refactoring to write units :)

So I only ask to be pragmatic

File: src/custom-apps/views/CustomAppList.tsx (hunk 2 above)
Comment (orzechdev) [reply src/custom-apps/views/CustomAppList.tsx]: I have extracted it into `findById` util.

**Comment by orzechdev** [comment]
@lkostrowski @Droniu @andrzejewsky 

I have added:
- objects with resolvers for custom apps urls,
- `findById` util,
- some unit tests for utils

Updated:
- icon for "Webhooks & Events" card on configuration page

**Review by lkostrowski — APPROVED** [review]
LGTM

In next PRs I think we should unify naming (local apps / webhooks & events)

**Review by andrzejewsky — APPROVED** [review]

**Comment by szczecha** [comment]
@orzechdev  @karolm-saleor  
Please make sure before approving and merging this ticket that e2e tests (apps.js) will work. Right now 3 of them fail probably due to visiting the new URL. 
image
Screenshot 2022-12-14 at 18 35 35

**Comment by karolm-saleor** [comment]
✓ QA Passed

**Review by karolm-saleor — APPROVED** [review]

**Review by lkostrowski — APPROVED** [review]

**Pull request merged**